src/libraries/System.Linq/src/System/Linq/OrderBy.cs

```diff
@@ -143,6 +169,10 @@ public static partial class Enumerable
             {
                 ThrowHelper.ThrowArgumentNullException(ExceptionArgument.source);
             }
+            if (keySelector is null)
```

Comment:
Are you sure? IOrderedEnumerable<out TElement>.CreateOrderedEnumerable<TKey>(...) is part of public API, so moving this check there would cause a behavior change.

Reply:
That's precisely why we should move the null check in the `CreateOrderedEnmerable` method. Its implementation calls the constructor:

https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Linq/src/System/Linq/OrderedEnumerable.cs#L72-L73

and the constructor no longer checks for null. As it stands the current PR will change null argument handling for that method.

Would it be possible to double check if we haven't accidentally moved null argument validation in other APIs?

Reply:
@eiriktsarpalis I think I will take tomorrow a look.

Reply:
https://github.com/dotnet/runtime/blob/57bfe474518ab5b7cfe6bf7424a79ce3af9d6657/src/libraries/System.Linq/src/System/Linq/OrderedEnumerable.cs#L72-L73

We could add here a check for `keySelector` and eliminate these checks in ThenBy* but we are also directly calling the ctor in Order*, OrderBy*.
`_source` should never be null here.

Reply:
Up to you. You could either try reverting the recommendation I made in https://github.com/dotnet/runtime/pull/70525#discussion_r896472688 or try something else. But could you add tests validating null arguments in CreateOrderedEnumerable?